### PR TITLE
*: enhance the table regions HTTP API

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

This API is used to get the table regions http://localhost:10080/tables/{DB}/{Table}/regions

In current implementation,  `getRegionsMeta` is used to get the region information.
That function first gets the regions in the key range, and then call `GetRegionByID` one by one for each region id.

The problem is that when the table is huge, it contains so many regions, calling the API is very slow because current implementation visits PD for each region.

### What is changed and how it works?

Use `ScanRegions` API to implement http://localhost:10080/tables/{DB}/{Table}/regions
ScanRegions accepts a key range and returns the region information, we need just one call to get the result instead of calling `GetRegionByID` for every region.

See also https://github.com/pingcap/pd/pull/1700 https://github.com/pingcap/pd/issues/1699

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

I test it manually on TiKV.

```
CREATE TABLE IF NOT EXISTS TEST_HOTSPOT(
      id                   BIGINT PRIMARY KEY,
      age                INT,
      user_name  VARCHAR(32),
      email 	 VARCHAR(128)
)
```

Split the table to get more regions:

```
SPLIT TABLE TEST_HOTSPOT BETWEEN (0) AND (9223372036854775807) REGIONS 128;
```

See the table regions:

```
mysql> show table TEST_HOTSPOT regions;
+-----------+----------------------------+----------------------------+-----------+-----------------+-------+------------+
| REGION_ID | START_KEY                  | END_KEY                    | LEADER_ID | LEADER_STORE_ID | PEERS | SCATTERING |
+-----------+----------------------------+----------------------------+-----------+-----------------+-------+------------+
|        48 | t_47_                      | t_47_r_72057594037927935   |        49 |               1 | 49    |          0 |
|        50 | t_47_r_72057594037927935   | t_47_r_144115188075855870  |        51 |               1 | 51    |          0 |
|        52 | t_47_r_144115188075855870  | t_47_r_216172782113783805  |        53 |               1 | 53    |          0 |
|        54 | t_47_r_216172782113783805  | t_47_r_288230376151711740  |        55 |               1 | 55    |          0 |
|        56 | t_47_r_288230376151711740  | t_47_r_360287970189639675  |        57 |               1 | 57    |          0 |
|        58 | t_47_r_360287970189639675  | t_47_r_432345564227567610  |        59 |               1 | 59    |          0 |
|        60 | t_47_r_432345564227567610  | t_47_r_504403158265495545  |        61 |               1 | 61    |          0 |
|        62 | t_47_r_504403158265495545  | t_47_r_576460752303423480  |        63 |               1 | 63    |          0 |
|        64 | t_47_r_576460752303423480  | t_47_r_648518346341351415  |        65 |               1 | 65    |          0 |
|        66 | t_47_r_648518346341351415  | t_47_r_720575940379279350  |        67 |               1 | 67    |          0 |
|        68 | t_47_r_720575940379279350  | t_47_r_792633534417207285  |        69 |               1 | 69    |          0 |
|        70 | t_47_r_792633534417207285  | t_47_r_864691128455135220  |        71 |               1 | 71    |          0 |
|        72 | t_47_r_864691128455135220  | t_47_r_936748722493063155  |        73 |               1 | 73    |          0 |
|        74 | t_47_r_936748722493063155  | t_47_r_1008806316530991090 |        75 |               1 | 75    |          0 |
|        76 | t_47_r_1008806316530991090 | t_47_r_1080863910568919025 |        77 |               1 | 77    |          0 |
|         2 | t_47_r_1080863910568919025 |                            |         3 |               1 | 3     |          0 |
+-----------+----------------------------+----------------------------+-----------+-----------------+-------+------------+
16 rows in set (0.03 sec)
```

Compare it with the result from the HTTP API:

```
{
 "name": "TEST_HOTSPOT",
 "id": 47,
 "record_regions": [
  {
   "region_id": 48,
   "leader": {
    "id": 49,
    "store_id": 1
   },
   "peers": [
    {
     "id": 49,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 24
   }
  },
  {
   "region_id": 50,
   "leader": {
    "id": 51,
    "store_id": 1
   },
   "peers": [
    {
     "id": 51,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 25
   }
  },
  {
   "region_id": 52,
   "leader": {
    "id": 53,
    "store_id": 1
   },
   "peers": [
    {
     "id": 53,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 26
   }
  },
  {
   "region_id": 54,
   "leader": {
    "id": 55,
    "store_id": 1
   },
   "peers": [
    {
     "id": 55,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 27
   }
  },
  {
   "region_id": 56,
   "leader": {
    "id": 57,
    "store_id": 1
   },
   "peers": [
    {
     "id": 57,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 28
   }
  },
  {
   "region_id": 58,
   "leader": {
    "id": 59,
    "store_id": 1
   },
   "peers": [
    {
     "id": 59,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 29
   }
  },
  {
   "region_id": 60,
   "leader": {
    "id": 61,
    "store_id": 1
   },
   "peers": [
    {
     "id": 61,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 30
   }
  },
  {
   "region_id": 62,
   "leader": {
    "id": 63,
    "store_id": 1
   },
   "peers": [
    {
     "id": 63,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 31
   }
  },
  {
   "region_id": 64,
   "leader": {
    "id": 65,
    "store_id": 1
   },
   "peers": [
    {
     "id": 65,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 32
   }
  },
  {
   "region_id": 66,
   "leader": {
    "id": 67,
    "store_id": 1
   },
   "peers": [
    {
     "id": 67,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 33
   }
  },
  {
   "region_id": 68,
   "leader": {
    "id": 69,
    "store_id": 1
   },
   "peers": [
    {
     "id": 69,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 34
   }
  },
  {
   "region_id": 70,
   "leader": {
    "id": 71,
    "store_id": 1
   },
   "peers": [
    {
     "id": 71,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 35
   }
  },
  {
   "region_id": 72,
   "leader": {
    "id": 73,
    "store_id": 1
   },
   "peers": [
    {
     "id": 73,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 36
   }
  },
  {
   "region_id": 74,
   "leader": {
    "id": 75,
    "store_id": 1
   },
   "peers": [
    {
     "id": 75,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 37
   }
  },
  {
   "region_id": 76,
   "leader": {
    "id": 77,
    "store_id": 1
   },
   "peers": [
    {
     "id": 77,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 38
   }
  },
  {
   "region_id": 2,
   "leader": {
    "id": 3,
    "store_id": 1
   },
   "peers": [
    {
     "id": 3,
     "store_id": 1
    }
   ],
   "region_epoch": {
    "conf_ver": 1,
    "version": 38
   }
  }
 ],
 "indices": [
  {
   "name": "i_email",
   "id": 1,
   "regions": [
    {
     "region_id": 48,
     "leader": {
      "id": 49,
      "store_id": 1
     },
     "peers": [
      {
       "id": 49,
       "store_id": 1
      }
     ],
     "region_epoch": {
      "conf_ver": 1,
      "version": 24
     }
    }
   ]
  }
 ]
}
```

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -152,24 +152,22 @@ github.com/pingcap/check v0.0.0-20190102082844-67f458068fc8 h1:USx2/E1bX46VG32FI
...
 github.com/pingcap/kvproto v0.0.0-20190910074005-0e61b6f435c1/go.mod h1:QMdbTAXCHzzygQzqcG9uVUgU2fKeSN1GmfMiykdSzzY=
-github.com/pingcap/log v0.0.0-20190214045112-b37da76f67a7/go.mod h1:xsfkWVaFVV5B8e1K9seWfyJWFrIhbtUTAD8NV1Pq3+w=
 github.com/pingcap/log v0.0.0-20190715063458-479153f07ebd h1:hWDol43WY5PGhsh3+8794bFHY1bPrmu6bTalpssCrGg=
 github.com/pingcap/log v0.0.0-20190715063458-479153f07ebd/go.mod h1:WpHUKhNZ18v116SvGrmjkA9CBhYmuUTKL+p8JC9ANEw=
 github.com/pingcap/parser v0.0.0-20190923031704-33636bc5e5d6 h1:PyjsTUD8gJ6QGilbwiy/TTn89J84/69Pj9LixOd/fFE=
 github.com/pingcap/parser v0.0.0-20190923031704-33636bc5e5d6/go.mod h1:1FNvfp9+J0wvc4kl8eGNh7Rqrxveg15jJoWo/a0uHwA=
-github.com/pingcap/pd v0.0.0-20190712044914-75a1f9f3062b h1:oS9PftxQqgcRouKhhdaB52tXhVLEP7Ng3Qqsd6Z18iY=
-github.com/pingcap/pd v0.0.0-20190712044914-75a1f9f3062b/go.mod h1:3DlDlFT7EF64A1bmb/tulZb6wbPSagm5G4p1AlhaEDs=
+github.com/pingcap/pd v1.1.0-beta.0.20190923032047-5c648dc365e0 h1:GIEq+wZfrl2bcJxpuSrEH4H7/nlf5YdmpS+dU9lNIt8=
+github.com/pingcap/pd v1.1.0-beta.0.20190923032047-5c648dc365e0/go.mod h1:G/6rJpnYwM0LKMec2rI82/5Kg6GaZMvlfB+e6/tvYmI=
```
Comment (disksing) [inline go.sum]: weird, any idea why it changes to `v1.1.0`?

File: server/http_handler.go [hunk 2]
```diff
@@ -1010,14 +1020,22 @@ func (h tableHandler) getRegionsByID(tbl table.Table, id int64, name string) (*T
 		indices[i].Name = index.Meta().Name.String()
 		indices[i].ID = indexID
 		startKey, endKey := tablecodec.GetTableIndexKeyRange(id, indexID)
-		rIDs, err := h.RegionCache.ListRegionIDsInKeyRange(tikv.NewBackoffer(context.Background(), 500), startKey, endKey)
+		regions, peers, err := pdCli.ScanRegions(ctx, startKey, endKey, 0)
```
Comment (disksing) [inline server/http_handler.go]: Could you unify the limit to 1 or 0?

File: store/tikv/store_test.go [hunk 3]
```diff
@@ -168,14 +168,14 @@ func (c *mockPDClient) GetRegionByID(ctx context.Context, regionID uint64) (*met
...
-func (c *mockPDClient) ScanRegions(ctx context.Context, startKey []byte, limit int) ([]*metapb.Region, []*metapb.Peer, error) {
+func (c *mockPDClient) ScanRegions(ctx context.Context, startKey []byte, endKey []byte, limit int) ([]*metapb.Region, []*metapb.Peer, error) {
 	c.RLock()
 	defer c.RUnlock()
 
 	if c.stop {
 		return nil, nil, errors.Trace(errStopped)
 	}
-	return c.client.ScanRegions(ctx, startKey, limit)
+	return c.client.ScanRegions(ctx, startKey, nil, limit)
```
Comment (disksing) [inline store/tikv/store_test.go]: pass `endKey` too?

File: go.sum (hunk 1 above)
Comment (tiancaiamao) [reply go.sum]: I use the latest pd in go.mod
https://github.com/pingcap/tidb/pull/12357/files#diff-37aff102a57d3d7b797f152915a6dc16R44

File: server/http_handler.go (hunk 2 above)
Comment (tiancaiamao) [reply server/http_handler.go]: -1 ?

Comment (disksing) [reply server/http_handler.go]: Yes. Either -1 or 0 is ok, but not use them both.

File: store/tikv/store_test.go (hunk 3 above)
Comment (tiancaiamao) [reply store/tikv/store_test.go]: done

File: server/http_handler.go (hunk 2 above)
Comment (tiancaiamao) [reply server/http_handler.go]: Now I change them to -1

File: store/mockstore/mocktikv/cluster.go [hunk 4]
```diff
@@ -285,14 +285,22 @@ func (c *Cluster) ScanRegions(key []byte, limit int) ([]*metapb.Region, []*metap
...
 			return true
 		}
-		return bytes.Compare(regions[i].Meta.GetEndKey(), key) > 0
+		return bytes.Compare(regions[i].Meta.GetEndKey(), startKey) > 0
 	})
 	regions = regions[keyLocation:]
-	if len(regions) > limit {
+	if len(endKey) > 0 {
+		var pos int
+		for pos = 1; pos < len(regions); pos++ {
```
Comment (disksing) [inline store/mockstore/mocktikv/cluster.go]: why not use binary search?

Comment (tiancaiamao) [reply store/mockstore/mocktikv/cluster.go]: Done.
Although I change it to binary search, I still don't think this change is meaningful.
There is a `sort.Slice` function, and it's a much heavy operation then search.

**Comment by tiancaiamao** [comment]
PTAL @lysu @jackysp

**Comment by tiancaiamao** [comment]
PTAL @lysu @jackysp

**Pull request merged**